opendevin/runtime/docker/ssh_box.py

```diff
@@ -221,7 +221,9 @@ def __init__(
                 raise Exception(
                     'Persistent sandbox is currently designed for opendevin user only. Please set run_as_devin=True in your config.toml'
                 )
-            self.instance_id = 'persisted'
+            self.instance_id = 'persisted ' + config.workspace_mount_path[1:].replace(
+                '/', '_'
+            )
```

Comment:
Revert changes in other files.
Replace 224-226 with this instead:
```py
            self.instance_id = 'persisted'
            if config.workspace_mount_path and len(config.workspace_mount_path) > 1:
                self.instance_id = ' '.join([self.instance_id, config.workspace_mount_path[1:].replace('/', '_')])
```

Reply:
docker name only supports alphanumeric [a-zA-Z0-9][a-zA-Z0-9_.-]+

Reply:
Ignore the outdated "replace" statement in my comment. My request would be to only apply changes to `ssh_box.py` to reach your goal.

Reply:
> `config.workspace_mount_path` will always have a value.

Why shouldn't the default value of `workspace_mount_path` be ''?

Reply:
I'm refusing to discuss obvious and unwanted introduction of behavorial code changes.

Reply:
There are no behavioral code changes.

Reply:
It needs to be an absolute path, doesn't it? So an empty string is not a valid value

Reply:
`None` also is not a valid one.

Reply:
Then unnecessary lines need to be added for mypy's sake.

> ```python
> if config.workspace_mount_path
> ```

This line conveys that it may have falsy value but that is not true.

Reply:
Empty string is a falsy value as well as None.

Reply:
The last time I removed None from workspace_mount_path, giving an actual default value, I introduced a pretty bad bug. I too wish it had a required, not optional, value, but optional was necessary to begin with. On the bright side,`finalize_config()` should _always_ be called when initializing config, so it will receive a valid value.

I agree it's not necessary to make it an invalid string.

Reply:
> This line conveys that it may have falsy values but that is not true.
> If the core has wrong values, it needs to be addressed.

How about replacing that line with
`assert config.workspace_mount_... is not None` ?

Or, if that's too much 😅, how about
```
if not config.workspace_mount...:
  # this will never happen (tm)
  # finalize_config should always make this a valid absolute path
  logger.debug("What should never happen, happened. Please report this as a bug.")

Reply:
[Initialize a string variable in Python: "" or None?](https://stackoverflow.com/q/1398164)

Reply:
> [Initialize a string variable in Python: "" or None?](https://stackoverflow.com/q/1398164)
> 
> But None is already an invalid one. So. I just changed it.

It's a hack to suit your needs that should be solved differently as my 1st comment showed.
The _initialization_ of the class should keep the value as None as the default.
Subsequent methods, like `finalize_config`, might take care of it, but that is intentionally.

You're mixing 2 different issues in one PR and should probably split it up in 2.

Reply:
> The _initialization_ of the class should keep the value as None as the default.

Why?

Reply:
I agree with @tobitege that it's an unnecessary change. Please note that the post linked doesn't support the change. The variable needs to be declared as an optional. Standard syntax for optional is `or None`. Default value empty string would be okay if it was a valid value, but it's not.

In general, IMHO a value that looks like it's a real value but it's not, is more error-prone down the road.

Can you please tell what problems you see with the potential alternatives I outlined?

Reply:
> > The _initialization_ of the class should keep the value as None as the default.
> 
> Why?

Look up "sentinel value" pattern.

Reply:
@enyst https://github.com/OpenDevin/OpenDevin/pull/2412#discussion_r1637806929

Reply:
> "sentinel value"

mypy does not care about that.

![image](https://github.com/OpenDevin/OpenDevin/assets/7231077/ff14ff87-e24f-4f9e-bb26-03067efbf752)

![image](https://github.com/OpenDevin/OpenDevin/assets/7231077/c1cd72b5-921f-4d1a-8477-6a98ba025e2d)

Reply:
@SmartManoj I answered exactly to that comment, and I suggested two alternatives. Can you please check the comments? https://github.com/OpenDevin/OpenDevin/pull/2412#discussion_r1637836107

Reply:
Missed that comment. Those are also for pleasuring `mypy` only. If it is so, how about `# type: ignore`? which conveys the type is wrong. So, changed the type directly.

Reply:
> [Initialize a string variable in Python: "" or None?](https://stackoverflow.com/q/1398164)
> 
> But None is already an invalid one. So. I just changed it.

Did you accept [this answer](https://stackoverflow.com/a/1398178)?

Reply:
I answered that comment too, above. Can you please take a look at a refreshed page? We talked a lot fast, it seems. It's normal that we're missing comments, but we do need to take the time to read, otherwise it gets a little difficult to understand the points made. 😓

Reply:
> Missed that comment. Those are also for pleasuring mypy only. If it is so, how about # type: ignore? which conveys the type is wrong.

If you prefer `#type ignore`? That works too.

Conveying that there is something wrong with the type, to the developer eyes, is exactly the point. If it does that job, good.

Reply:
> > "sentinel value"
> 
> @tobitege mypy does not care about that.

Mypy is not a developer and doesn't have to think about the concepts, patterns and best practices of the project.

Run `mypy` over `config.py` and you won't get any error.

If someone uses invald values, they will receive runtime errors. That is not a reason to change initialization of the class as it is on purpose this way.

Reply:
mypy should be run on the whole folder.